service/storage_service.cc

```diff
@@ -5880,6 +5886,87 @@ future<> storage_service::add_tablet_replica(table_id table, dht::token token, l
     });
 }
 
+future<> storage_service::del_tablet_replica(table_id table, dht::token token, locator::tablet_replica dst, loosen_constraints force) {
+    auto holder = _async_gate.hold();
+
+    if (this_shard_id() != 0) {
+        // group0 is only set on shard 0.
+        co_return co_await container().invoke_on(0, [&] (auto& ss) {
+            return ss.del_tablet_replica(table, token, dst, force);
+        });
+    }
+
+    while (true) {
+        auto guard = co_await _group0->client().start_operation(&_group0_as, raft_timeout{});
+
+        while (_topology_state_machine._topology.is_busy()) {
+            const auto tstate = *_topology_state_machine._topology.tstate;
+            if (tstate == topology::transition_state::tablet_draining ||
+                tstate == topology::transition_state::tablet_migration) {
```

Comment:
I know this code is used in other places and not introduced in PR. I want to know what is special about the topology::transition_state::tablet_draining and topology::transition_state::tablet_migration? Why do we break even if the topology is busy for the two states?

Reply:
Because when topology coordinator is busy with those states, it still can process transition request that's placed by this code later

Reply:
Yes, but as Tomek mentioned, if we are in tablet_draining, we should not overwrite it to tablet_migration directly. We should wait for tablet_draining to finish, just like other states in topology transition.

Reply:
Yes, I agree. The same stays true for move_tablet() and add_tablet_replica(). I'll get back to those as a followup

Reply:
> Yes, but as Tomek mentioned, if we are in tablet_draining, we should not overwrite it to tablet_migration directly. We should wait for tablet_draining to finish, just like other states in topology transition.

We don't have to wait for tablet_drain to finish, we could start a migration inside it. There's advantage to allowing that, because we can fix load imbalance earlier. We just need to preserve tablet_drain.

Reply:
But if there is a pending tablet_drain, will tablet_migration be executed at the same time? Shouldn't the tablet_drain be executed with higher priority, because we want to finish the node ops triggered  tablet_drain as soon as possible, e.g., we run decommission, we use tablet_drain to move tablets from it. Btw, what other higher level ops use tablet_drain?

Reply:
I would say that if admin invokes the API, that has at least as high priority as node operations, maybe higher, because it's the same actor's decision. If admin thinks otherwise, he should wait until decommission is finished before he tries the API.

Reply:
drain is also used by removenode.

Reply:
@tgrabiec  I still not sure the question above "But if there is a pending tablet_drain, will tablet_migration be executed at the same time?"  If the answer is no, adding the work before tablet_drain is done is useless, because tablet_migration work will not be executed even if it is added. 

The admin is not going to use the remove_replica directly. Mostly probably, the cql user which is not aware of the node ops like decommission or removenode will trigger the remove replica code in this PR using CQL. The cql user usually has now way to know the node ops neither. Imagine a scylla cloud user.

Reply:
> drain is also used by removenode.

Does replace uses tablet_drain track too?

I saw

```
    state replacing {
        rp_join_group0 : join_group0
        rp_left_token_ring : left_token_ring
        rp_tablet_draining : tablet_draining
        rp_write_both_read_old : write_both_read_old
        rp_write_both_read_new : write_both_read_new
        [*] --> rp_join_group0
        rp_join_group0 --> rp_left_token_ring: rollback
        rp_join_group0 --> rp_tablet_draining
        rp_tablet_draining --> rp_write_both_read_old
        rp_tablet_draining --> rp_left_token_ring: rollback
        rp_join_group0 --> [*]: rejected
        rp_write_both_read_old --> rp_write_both_read_new: streaming completed
        rp_write_both_read_old --> rp_left_token_ring: rollback 
        rp_write_both_read_new --> [*]
        rp_left_token_ring --> [*]
    }
```

Why do we use tablet_drain for replace?

Reply:
> @tgrabiec I still not sure the question above "But if there is a pending tablet_drain, will tablet_migration be executed at the same time?" If the answer is no, adding the work before tablet_drain is done is useless, because tablet_migration work will not be executed even if it is added.

The answer is yes, it will be executed concurrently. So you can migrate tablets immediately when we're in the middle of draining some node. I think that's good, because we're more responsive. Maybe not very important either, but since it's a single if condition, it's better to allow it.  

> 
> The admin is not going to use the remove_replica directly. Mostly probably, the cql user which is not aware of the node ops like decommission or removenode will trigger the remove replica code in this PR using CQL. The cql user usually has now way to know the node ops neither. Imagine a scylla cloud user.

I think ALTER KEYSPACE will not use this API but make updates directly, and it excludes with any other topology change, so this problem of busy topology does not apply.

Reply:
> > drain is also used by removenode.
> 
> Does replace uses tablet_drain track too?

No, for replace we rebuild in the background in the regular tablet migration track.
The state machine still goes through tablet drain state because the flow is statically defined. It will do nothing for replace.